src/cargo/ops/cargo_new.rs

```diff
@@ -541,16 +613,127 @@ fn global_config(config: &Config) -> CargoResult<CargoNewConfig> {
+///    http://doc.rust-lang.org/std/fs/fn.read_dir.html
+fn walk_template_dir(dir: &Path, cb: &mut FnMut(DirEntry) -> CargoResult<()>) -> CargoResult<()> {
+    let attr = try!(fs::metadata(&dir));
+    let ignore_files = vec![".gitignore"];
+
+    if attr.is_dir() {
+        for entry in try!(fs::read_dir(dir)) {
+            let entry = try!(entry);
+            let attr = try!(fs::metadata(&entry.path()));
+            if attr.is_dir() {
+                if !&entry.path().to_str().unwrap().contains(".git") {
+                    try!(walk_template_dir(&entry.path(), cb));
+                }
+            } else {
+                if let Some(file_name) = entry.path().file_name() {
+                    if ignore_files.contains(&file_name.to_str().unwrap()) {
+                        continue
+                    }
+                }
+                try!(cb(entry));
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Create a generic template
+///
+/// This consists of a Cargo.toml, and a src directory.
+fn create_generic_template() -> Vec<Box<TemplateFile>> {
+    let template_file = Box::new(InMemoryTemplateFile::new(PathBuf::from("Cargo.toml"),
+    String::from(r#"[package]
+name = "{{name}}"
+version = "0.1.0"
+authors = [{{toml-escape author}}]
+"#)));
+    vec![template_file]
+}
+
+/// Create a new "lib" project
+fn create_lib_template() -> Vec<Box<TemplateFile>> {
+    let mut template_files = create_generic_template();
+    let lib_file = Box::new(InMemoryTemplateFile::new(PathBuf::from("src/lib.rs"),
+    String::from(r#"#[test]
+fn it_works() {
+}
+"#)));
+    template_files.push(lib_file);
+    template_files
+}
+
+/// Create a new "bin" project
+fn create_bin_template() -> Vec<Box<TemplateFile>> {
+    let mut template_files = create_generic_template();
+    let main_file = Box::new(InMemoryTemplateFile::new(PathBuf::from("src/main.rs"),
+String::from("fn main() {
+    println!(\"Hello, world!\");
+}
+")));
+    template_files.push(main_file);
+    template_files
+}
+
+// Create subdirs for files deeper than <project-name>/<src>.
+// 'path' could refers to a file inside a directory which doesn't yet exist
+// to figure out if this is the case, get all the components in the dest_file_name and check
+// how many there are. Files in the root of the new project direcotory will have two
+// components, anything more than that means the file is in a sub-directory, so we need
+// to create it.
+fn create_deep_subdirs(path: &Path) {
+    let components  = path.components().collect::<Vec<_>>();
+    if components.len() > 2 {
+        if let Some(p) = path.parent() {
+            if fs::metadata(&p).is_err() {
+                let _ = fs::create_dir_all(&p);
+            }
+        }
+    }
+}
+
+fn abs_to_rel_path(abspath: &Path, rel_part: &Path) -> CargoResult<PathBuf> {
+    let abspath_str= try!(abspath.to_str().ok_or(
+            human(format!("Could not convert path to string: {}", abspath.display()))));
```

Comment:
Stylistically Cargo error messages typically start with lowercase letters (here and a few other locations too)